// @tldraw/tlschema/dist-esm/assets/TLVideoAsset.mjs
import { defineMigrations } from "@tldraw/store";
import { T } from "@tldraw/validate";
import { createAssetValidator } from "./TLBaseAsset.mjs";
const videoAssetValidator = createAssetValidator(
  "video",
  T.object({
    w: T.number,
    h: T.number,
    name: T.string,
    isAnimated: T.boolean,
    mimeType: T.string.nullable(),
    src: T.srcUrl.nullable()
  })
);
const Versions = {
  AddIsAnimated: 1,
  RenameWidthHeight: 2,
  MakeUrlsValid: 3
};
const videoAssetMigrations = defineMigrations({
  currentVersion: Versions.MakeUrlsValid,
  migrators: {
    [Versions.AddIsAnimated]: {
      up: (asset) => {
        return {
          ...asset,
          props: {
            ...asset.props,
            isAnimated: false
          }
        };
      },
      down: (asset) => {
        const { isAnimated, ...rest } = asset.props;
        return {
          ...asset,
          props: rest
        };
      }
    },
    [Versions.RenameWidthHeight]: {
      up: (asset) => {
        const { width, height, ...others } = asset.props;
        return { ...asset, props: { w: width, h: height, ...others } };
      },
      down: (asset) => {
        const { w, h, ...others } = asset.props;
        return { ...asset, props: { width: w, height: h, ...others } };
      }
    },
    [Versions.MakeUrlsValid]: {
      up: (asset) => {
        const src = asset.props.src;
        if (src && !T.srcUrl.isValid(src)) {
          return { ...asset, props: { ...asset.props, src: "" } };
        }
        return asset;
      },
      down: (asset) => asset
    }
  }
});
export {
  videoAssetMigrations,
  videoAssetValidator
};
