// @tldraw/editor/dist-esm/lib/config/TLUserPreferences.mjs
import { atom } from "@tldraw/state";
import { defineMigrations, migrate } from "@tldraw/store";
import { getDefaultTranslationLocale } from "@tldraw/tlschema";
import { T } from "@tldraw/validate";
import { uniqueId } from "../utils/uniqueId.mjs";
const USER_DATA_KEY = "TLDRAW_USER_DATA_v3";
const userTypeValidator = T.object({
  id: T.string,
  name: T.string.nullable().optional(),
  locale: T.string.nullable().optional(),
  color: T.string.nullable().optional(),
  isDarkMode: T.boolean.nullable().optional(),
  animationSpeed: T.number.nullable().optional(),
  edgeScrollSpeed: T.number.nullable().optional(),
  isSnapMode: T.boolean.nullable().optional(),
  isWrapMode: T.boolean.nullable().optional()
});
const Versions = {
  AddAnimationSpeed: 1,
  AddIsSnapMode: 2,
  MakeFieldsNullable: 3,
  AddEdgeScrollSpeed: 4,
  AddExcalidrawSelectMode: 5
};
const userMigrations = defineMigrations({
  currentVersion: Versions.AddExcalidrawSelectMode,
  migrators: {
    [Versions.AddAnimationSpeed]: {
      up: (user) => {
        return {
          ...user,
          animationSpeed: 1
        };
      },
      down: ({ animationSpeed: _, ...user }) => {
        return user;
      }
    },
    [Versions.AddIsSnapMode]: {
      up: (user) => {
        return { ...user, isSnapMode: false };
      },
      down: ({ isSnapMode: _, ...user }) => {
        return user;
      }
    },
    [Versions.MakeFieldsNullable]: {
      up: (user) => {
        return user;
      },
      down: (user) => {
        return {
          id: user.id,
          name: user.name ?? defaultUserPreferences.name,
          locale: user.locale ?? defaultUserPreferences.locale,
          color: user.color ?? defaultUserPreferences.color,
          animationSpeed: user.animationSpeed ?? defaultUserPreferences.animationSpeed,
          isDarkMode: user.isDarkMode ?? defaultUserPreferences.isDarkMode,
          isSnapMode: user.isSnapMode ?? defaultUserPreferences.isSnapMode,
          isWrapMode: user.isWrapMode ?? defaultUserPreferences.isWrapMode
        };
      }
    },
    [Versions.AddEdgeScrollSpeed]: {
      up: (user) => {
        return {
          ...user,
          edgeScrollSpeed: 1
        };
      },
      down: ({ edgeScrollSpeed: _, ...user }) => {
        return user;
      }
    },
    [Versions.AddExcalidrawSelectMode]: {
      up: (user) => {
        return { ...user, isWrapMode: false };
      },
      down: ({ isWrapMode: _, ...user }) => {
        return user;
      }
    }
  }
});
const USER_COLORS = [
  "#FF802B",
  "#EC5E41",
  "#F2555A",
  "#F04F88",
  "#E34BA9",
  "#BD54C6",
  "#9D5BD2",
  "#7B66DC",
  "#02B1CC",
  "#11B3A3",
  "#39B178",
  "#55B467"
];
function getRandomColor() {
  return USER_COLORS[Math.floor(Math.random() * USER_COLORS.length)];
}
function userPrefersDarkUI() {
  if (typeof window === "undefined") {
    return false;
  }
  return window.matchMedia?.("(prefers-color-scheme: dark)")?.matches ?? false;
}
function userPrefersReducedMotion() {
  if (typeof window === "undefined") {
    return false;
  }
  return window.matchMedia?.("(prefers-reduced-motion: reduce)")?.matches ?? false;
}
const defaultUserPreferences = Object.freeze({
  name: "New User",
  locale: getDefaultTranslationLocale(),
  color: getRandomColor(),
  isDarkMode: false,
  edgeScrollSpeed: 1,
  animationSpeed: userPrefersReducedMotion() ? 0 : 1,
  isSnapMode: false,
  isWrapMode: false
});
function getFreshUserPreferences() {
  return {
    id: uniqueId()
  };
}
function migrateUserPreferences(userData) {
  if (userData === null || typeof userData !== "object") {
    return getFreshUserPreferences();
  }
  if (!("version" in userData) || !("user" in userData) || typeof userData.version !== "number") {
    return getFreshUserPreferences();
  }
  const migrationResult = migrate({
    value: userData.user,
    fromVersion: userData.version,
    toVersion: userMigrations.currentVersion ?? 0,
    migrations: userMigrations
  });
  if (migrationResult.type === "error") {
    return getFreshUserPreferences();
  }
  try {
    userTypeValidator.validate(migrationResult.value);
  } catch (e) {
    return getFreshUserPreferences();
  }
  return migrationResult.value;
}
function loadUserPreferences() {
  const userData = typeof window === "undefined" ? null : JSON.parse(window?.localStorage?.getItem(USER_DATA_KEY) || "null") ?? null;
  return migrateUserPreferences(userData);
}
const globalUserPreferences = atom("globalUserData", null);
function storeUserPreferences() {
  if (typeof window !== "undefined" && window.localStorage) {
    window.localStorage.setItem(
      USER_DATA_KEY,
      JSON.stringify({
        version: userMigrations.currentVersion,
        user: globalUserPreferences.get()
      })
    );
  }
}
function setUserPreferences(user) {
  userTypeValidator.validate(user);
  globalUserPreferences.set(user);
  storeUserPreferences();
  broadcastUserPreferencesChange();
}
const isTest = typeof process !== "undefined" && process.env.NODE_ENV === "test";
const channel = typeof BroadcastChannel !== "undefined" && !isTest ? new BroadcastChannel("tldraw-user-sync") : null;
channel?.addEventListener("message", (e) => {
  const data = e.data;
  if (data?.type === broadcastEventKey && data?.origin !== getBroadcastOrigin()) {
    globalUserPreferences.set(migrateUserPreferences(data.data));
  }
});
let _broadcastOrigin = null;
function getBroadcastOrigin() {
  if (_broadcastOrigin === null) {
    _broadcastOrigin = uniqueId();
  }
  return _broadcastOrigin;
}
const broadcastEventKey = "tldraw-user-preferences-change";
function broadcastUserPreferencesChange() {
  channel?.postMessage({
    type: broadcastEventKey,
    origin: getBroadcastOrigin(),
    data: {
      user: getUserPreferences(),
      version: userMigrations.currentVersion
    }
  });
}
function getUserPreferences() {
  let prefs = globalUserPreferences.get();
  if (!prefs) {
    prefs = loadUserPreferences();
    globalUserPreferences.set(prefs);
  }
  return prefs;
}
export {
  USER_COLORS,
  defaultUserPreferences,
  getFreshUserPreferences,
  getUserPreferences,
  setUserPreferences,
  userPrefersDarkUI,
  userPrefersReducedMotion
};
