// @tldraw/state/dist-cjs/lib/core/transactions.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transactions_exports = {};
__export(transactions_exports, {
  advanceGlobalEpoch: () => advanceGlobalEpoch,
  atomDidChange: () => atomDidChange,
  getGlobalEpoch: () => getGlobalEpoch,
  transact: () => transact,
  transaction: () => transaction
});
module.exports = __toCommonJS(transactions_exports);
var import_constants = require("./constants");
var import_EffectScheduler = require("./EffectScheduler");
var import_helpers = require("./helpers");
class Transaction {
  constructor(parent) {
    this.parent = parent;
  }
  initialAtomValues = /* @__PURE__ */ new Map();
  /**
   * Get whether this transaction is a root (no parents).
   *
   * @public
   */
  // eslint-disable-next-line no-restricted-syntax
  get isRoot() {
    return this.parent === null;
  }
  /**
   * Commit the transaction's changes.
   *
   * @public
   */
  commit() {
    if (this.isRoot) {
      const atoms = this.initialAtomValues;
      this.initialAtomValues = /* @__PURE__ */ new Map();
      flushChanges(atoms.keys());
    } else {
      this.initialAtomValues.forEach((value, atom) => {
        if (!this.parent.initialAtomValues.has(atom)) {
          this.parent.initialAtomValues.set(atom, value);
        }
      });
    }
  }
  /**
   * Abort the transaction.
   *
   * @public
   */
  abort() {
    inst.globalEpoch++;
    this.initialAtomValues.forEach((value, atom) => {
      atom.set(value);
      atom.historyBuffer?.clear();
    });
    this.commit();
  }
}
const inst = (0, import_helpers.singleton)("transactions", () => ({
  // The current epoch (global to all atoms).
  globalEpoch: import_constants.GLOBAL_START_EPOCH + 1,
  // Whether any transaction is reacting.
  globalIsReacting: false,
  currentTransaction: null
}));
function getGlobalEpoch() {
  return inst.globalEpoch;
}
function flushChanges(atoms) {
  if (inst.globalIsReacting) {
    throw new Error("cannot change atoms during reaction cycle");
  }
  try {
    inst.globalIsReacting = true;
    const reactors = /* @__PURE__ */ new Set();
    const traverse = (node) => {
      if (node.lastTraversedEpoch === inst.globalEpoch) {
        return;
      }
      node.lastTraversedEpoch = inst.globalEpoch;
      if (node instanceof import_EffectScheduler.EffectScheduler) {
        reactors.add(node);
      } else {
        ;
        node.children.visit(traverse);
      }
    };
    for (const atom of atoms) {
      atom.children.visit(traverse);
    }
    for (const r of reactors) {
      r.maybeScheduleEffect();
    }
  } finally {
    inst.globalIsReacting = false;
  }
}
function atomDidChange(atom, previousValue) {
  if (!inst.currentTransaction) {
    flushChanges([atom]);
  } else if (!inst.currentTransaction.initialAtomValues.has(atom)) {
    inst.currentTransaction.initialAtomValues.set(atom, previousValue);
  }
}
function advanceGlobalEpoch() {
  inst.globalEpoch++;
}
function transaction(fn) {
  const txn = new Transaction(inst.currentTransaction);
  inst.currentTransaction = txn;
  try {
    let rollback = false;
    const result = fn(() => rollback = true);
    if (rollback) {
      txn.abort();
    } else {
      txn.commit();
    }
    return result;
  } catch (e) {
    txn.abort();
    throw e;
  } finally {
    inst.currentTransaction = inst.currentTransaction.parent;
  }
}
function transact(fn) {
  if (inst.currentTransaction) {
    return fn();
  }
  return transaction(fn);
}
