// @tldraw/tlschema/dist-esm/shapes/TLBookmarkShape.mjs
import { defineMigrations } from "@tldraw/store";
import { T } from "@tldraw/validate";
import { assetIdValidator } from "../assets/TLBaseAsset.mjs";
const bookmarkShapeProps = {
  w: T.nonZeroNumber,
  h: T.nonZeroNumber,
  assetId: assetIdValidator.nullable(),
  url: T.linkUrl
};
const Versions = {
  NullAssetId: 1,
  MakeUrlsValid: 2
};
const bookmarkShapeMigrations = defineMigrations({
  currentVersion: Versions.MakeUrlsValid,
  migrators: {
    [Versions.NullAssetId]: {
      up: (shape) => {
        if (shape.props.assetId === void 0) {
          return { ...shape, props: { ...shape.props, assetId: null } };
        }
        return shape;
      },
      down: (shape) => {
        if (shape.props.assetId === null) {
          const { assetId: _, ...props } = shape.props;
          return { ...shape, props };
        }
        return shape;
      }
    },
    [Versions.MakeUrlsValid]: {
      up: (shape) => {
        const url = shape.props.url;
        if (url !== "" && !T.linkUrl.isValid(shape.props.url)) {
          return { ...shape, props: { ...shape.props, url: "" } };
        }
        return shape;
      },
      down: (shape) => shape
    }
  }
});
export {
  bookmarkShapeMigrations,
  bookmarkShapeProps
};
