// @tldraw/tlschema/dist-esm/shapes/TLVideoShape.mjs
import { defineMigrations } from "@tldraw/store";
import { T } from "@tldraw/validate";
import { assetIdValidator } from "../assets/TLBaseAsset.mjs";
const videoShapeProps = {
  w: T.nonZeroNumber,
  h: T.nonZeroNumber,
  time: T.number,
  playing: T.boolean,
  url: T.linkUrl,
  assetId: assetIdValidator.nullable()
};
const Versions = {
  AddUrlProp: 1,
  MakeUrlsValid: 2
};
const videoShapeMigrations = defineMigrations({
  currentVersion: Versions.MakeUrlsValid,
  migrators: {
    [Versions.AddUrlProp]: {
      up: (shape) => {
        return { ...shape, props: { ...shape.props, url: "" } };
      },
      down: (shape) => {
        const { url: _, ...props } = shape.props;
        return { ...shape, props };
      }
    },
    [Versions.MakeUrlsValid]: {
      up: (shape) => {
        const url = shape.props.url;
        if (url !== "" && !T.linkUrl.isValid(shape.props.url)) {
          return { ...shape, props: { ...shape.props, url: "" } };
        }
        return shape;
      },
      down: (shape) => shape
    }
  }
});
export {
  videoShapeMigrations,
  videoShapeProps
};
