// tldraw/dist-esm/lib/ui/components/LanguageMenu.mjs
import { jsx } from "react/jsx-runtime";
import { track, useEditor } from "@tldraw/editor";
import { useUiEvents } from "../context/events.mjs";
import { useLanguages } from "../hooks/useTranslation/useLanguages.mjs";
import { TldrawUiMenuCheckboxItem } from "./primitives/menus/TldrawUiMenuCheckboxItem.mjs";
import { TldrawUiMenuGroup } from "./primitives/menus/TldrawUiMenuGroup.mjs";
import { TldrawUiMenuSubmenu } from "./primitives/menus/TldrawUiMenuSubmenu.mjs";
const LanguageMenu = track(function LanguageMenu2() {
  const editor = useEditor();
  const trackEvent = useUiEvents();
  const { languages, currentLanguage } = useLanguages();
  return /* @__PURE__ */ jsx(TldrawUiMenuSubmenu, { id: "help menu language", label: "menu.language", children: /* @__PURE__ */ jsx(TldrawUiMenuGroup, { id: "languages", children: languages.map(({ locale, label }) => /* @__PURE__ */ jsx(
    TldrawUiMenuCheckboxItem,
    {
      id: `language-${locale}`,
      title: locale,
      label,
      checked: locale === currentLanguage,
      onSelect: () => {
        editor.user.updateUserPreferences({ locale });
        trackEvent("change-language", { source: "menu", locale });
      }
    },
    locale
  )) }) });
});
export {
  LanguageMenu
};
